File: sdk/lx-trading-cpp/src/execution.cpp
// LX Trading SDK - Execution Algorithms Implementation

#include <lx/trading/execution.hpp>
#include <lx/trading/client.hpp>
#include <thread>

namespace lx::trading {

// =============================================================================
// TWAP Executor
// =============================================================================

TwapExecutor::TwapExecutor(
    Client& client,
    std::string symbol,
    Side side,
    Decimal total_quantity,
    std::chrono::seconds duration,
    int num_slices)
    : client_(client),
      symbol_(std::move(symbol)),
      side_(side),
      total_quantity_(total_quantity),
      duration_(duration),
      num_slices_(num_slices) {}

std::future<ExecutionResult> TwapExecutor::execute() {
    return std::async(std::launch::async, [this]() {
        ExecutionResult result;
        result.start_time = now_ms();
        result.total_quantity = total_quantity_;
        result.completed = false;

        Decimal slice_qty = total_quantity_ / Decimal::from_double(num_slices_);
        auto interval = std::chrono::milliseconds(
            duration_.count() * 1000 / num_slices_);

        Decimal total_value;

        for (int i = 0; i < num_slices_; ++i) {
            if (cancelled_.load(std::memory_order_acquire)) {
                result.error = "Cancelled";
                break;
            }

            Decimal remaining = total_quantity_ - (slice_qty * Decimal::from_double(i));
            Decimal qty = (remaining < slice_qty) ? remaining : slice_qty;

            if (qty <= Decimal::zero()) break;

            try {
                Order order;
                if (side_ == Side::Buy) {
                    order = client_.buy(symbol_, qty).get();
                } else {
                    order = client_.sell(symbol_, qty).get();
                }

                result.orders.push_back(order);
                result.total_filled = result.total_filled + order.filled_quantity;

                if (order.average_price) {
                    total_value = total_value +
                        (order.filled_quantity * *order.average_price);
                }

                if (callback_) {
                    callback_(order, remaining - qty);
                }
            } catch (const std::exception& e) {
                result.error = e.what();
                break;
            }

            if (i < num_slices_ - 1) {
                std::this_thread::sleep_for(interval);
            }
        }

        result.end_time = now_ms();
        result.completed = result.error.empty();

        if (result.total_filled.is_positive()) {
            result.average_price = total_value / result.total_filled;
        }

        return result;
    });
}

void TwapExecutor::cancel() {
    cancelled_.store(true, std::memory_order_release);
}

// =============================================================================
// VWAP Executor
// =============================================================================

VwapExecutor::VwapExecutor(
    Client& client,
    std::string symbol,
    Side side,
    Decimal total_quantity,
    Decimal participation_rate,
    std::chrono::seconds max_duration)
    : client_(client),
      symbol_(std::move(symbol)),
      side_(side),
      total_quantity_(total_quantity),
      participation_rate_(participation_rate),
      max_duration_(max_duration) {}

std::future<ExecutionResult> VwapExecutor::execute() {
    return std::async(std::launch::async, [this]() {
        ExecutionResult result;
        result.start_time = now_ms();
        result.total_quantity = total_quantity_;
        result.completed = false;

        Decimal remaining = total_quantity_;
        const int check_interval_ms = 5000;
        int elapsed_ms = 0;
        Decimal total_value;

        while (remaining > Decimal::zero() &&
               elapsed_ms < max_duration_.count() * 1000) {
            if (cancelled_.load(std::memory_order_acquire)) {
                result.error = "Cancelled";
                break;
            }

            try {
                auto ticker = client_.ticker(symbol_).get();
                Decimal volume = ticker.volume_24h.value_or(Decimal::from_double(1000));

                // Calculate slice based on participation
                Decimal hourly_volume = volume / Decimal::from_double(24);
                Decimal slice_volume = hourly_volume * participation_rate_ /
                    Decimal::from_double(3600000.0 / check_interval_ms);
                Decimal qty = (remaining < slice_volume) ? remaining : slice_volume;

                if (qty > Decimal::zero()) {
                    Order order;
                    if (side_ == Side::Buy) {
                        order = client_.buy(symbol_, qty).get();
                    } else {
                        order = client_.sell(symbol_, qty).get();
                    }

                    result.orders.push_back(order);
                    result.total_filled = result.total_filled + order.filled_quantity;
                    remaining = remaining - order.filled_quantity;

                    if (order.average_price) {
                        total_value = total_value +
                            (order.filled_quantity * *order.average_price);
                    }

                    if (callback_) {
                        callback_(order, remaining);
                    }
                }
            } catch (const std::exception& e) {
                // Log but continue
            }

            std::this_thread::sleep_for(std::chrono::milliseconds(check_interval_ms));
            elapsed_ms += check_interval_ms;
        }

        result.end_time = now_ms();
        result.completed = result.error.empty() && remaining <= Decimal::zero();

        if (result.total_filled.is_positive()) {
            result.average_price = total_value / result.total_filled;
        }

        return result;
    });
}

void VwapExecutor::cancel() {
    cancelled_.store(true, std::memory_order_release);
}

// =============================================================================
// Iceberg Executor
// =============================================================================

IcebergExecutor::IcebergExecutor(
    Client& client,
    std::string symbol,
    Side side,
    Decimal total_quantity,
    Decimal visible_quantity,
    Decimal price,
    std::optional<std::string> venue)
    : client_(client),
      symbol_(std::move(symbol)),
      side_(side),
      total_quantity_(total_quantity),
      visible_quantity_(visible_quantity),
      price_(price),
      venue_(std::move(venue)) {}

std::future<ExecutionResult> IcebergExecutor::execute() {
    return std::async(std::launch::async, [this]() {
        ExecutionResult result;
        result.start_time = now_ms();
        result.total_quantity = total_quantity_;
        result.completed = false;

        Decimal remaining = total_quantity_;
        Decimal total_value;

        while (remaining > Decimal::zero()) {
            if (cancelled_.load(std::memory_order_acquire)) {
                result.error = "Cancelled";
                break;
            }

            Decimal qty = (remaining < visible_quantity_) ? remaining : visible_quantity_;

            try {
                Order order;
                if (side_ == Side::Buy) {
                    order = client_.limit_buy(
                        symbol_, qty, price_,
                        venue_ ? std::optional<std::string_view>(*venue_)
                               : std::nullopt).get();
                } else {
                    order = client_.limit_sell(
                        symbol_, qty, price_,
                        venue_ ? std::optional<std::string_view>(*venue_)
                               : std::nullopt).get();
                }

                // Wait for fill (simplified - in production would poll/stream)
                std::this_thread::sleep_for(std::chrono::milliseconds(500));

                result.orders.push_back(order);
                result.total_filled = result.total_filled + order.filled_quantity;
                remaining = remaining - order.filled_quantity;

                if (order.average_price) {
                    total_value = total_value +
                        (order.filled_quantity * *order.average_price);
                }

                if (callback_) {
                    callback_(order, remaining);
                }
            } catch (const std::exception& e) {
                result.error = e.what();
                break;
            }
        }

        result.end_time = now_ms();
        result.completed = result.error.empty() && remaining <= Decimal::zero();

        if (result.total_filled.is_positive()) {
            result.average_price = total_value / result.total_filled;
        }

        return result;
    });
}

void IcebergExecutor::cancel() {
    cancelled_.store(true, std::memory_order_release);
}

// =============================================================================
// Sniper Executor
// =============================================================================

SniperExecutor::SniperExecutor(
    Client& client,
    std::string symbol,
    Side side,
    Decimal quantity,
    Decimal target_price,
    std::chrono::seconds timeout)
    : client_(client),
      symbol_(std::move(symbol)),
      side_(side),
      quantity_(quantity),
      target_price_(target_price),
      timeout_(timeout) {}

std::future<ExecutionResult> SniperExecutor::execute() {
    return std::async(std::launch::async, [this]() {
        ExecutionResult result;
        result.start_time = now_ms();
        result.total_quantity = quantity_;
        result.completed = false;

        const int check_interval_ms = 100;
        int elapsed_ms = 0;

        while (elapsed_ms < timeout_.count() * 1000) {
            if (cancelled_.load(std::memory_order_acquire)) {
                result.error = "Cancelled";
                break;
            }

            try {
                auto ticker = client_.ticker(symbol_).get();

                bool should_execute = false;
                if (side_ == Side::Buy) {
                    if (ticker.ask && *ticker.ask <= target_price_) {
                        should_execute = true;
                    }
                } else {
                    if (ticker.bid && *ticker.bid >= target_price_) {
                        should_execute = true;
                    }
                }

                if (should_execute) {
                    Order order;
                    if (side_ == Side::Buy) {
                        order = client_.buy(symbol_, quantity_).get();
                    } else {
                        order = client_.sell(symbol_, quantity_).get();
                    }

                    result.orders.push_back(order);
                    result.total_filled = order.filled_quantity;
                    result.average_price = order.average_price;
                    result.completed = true;
                    result.end_time = now_ms();

                    if (callback_) {
                        callback_(order, Decimal::zero());
                    }

                    return result;
                }
            } catch (const std::exception&) {
                // Continue waiting
            }

            std::this_thread::sleep_for(std::chrono::milliseconds(check_interval_ms));
            elapsed_ms += check_interval_ms;
        }

        result.end_time = now_ms();
        if (result.error.empty() && !result.completed) {
            result.error = "Timeout";
        }

        return result;
    });
}

void SniperExecutor::cancel() {
    cancelled_.store(true, std::memory_order_release);
}

// =============================================================================
// POV Executor
// =============================================================================

PovExecutor::PovExecutor(
    Client& client,
    std::string symbol,
    Side side,
    Decimal total_quantity,
    Decimal target_participation,
    std::chrono::seconds max_duration,
    std::optional<Decimal> price_limit)
    : client_(client),
      symbol_(std::move(symbol)),
      side_(side),
      total_quantity_(total_quantity),
      target_participation_(target_participation),
      max_duration_(max_duration),
      price_limit_(price_limit) {}

std::future<ExecutionResult> PovExecutor::execute() {
    return std::async(std::launch::async, [this]() {
        ExecutionResult result;
        result.start_time = now_ms();
        result.total_quantity = total_quantity_;
        result.completed = false;

        // Similar to VWAP but with stricter participation targeting
        Decimal remaining = total_quantity_;
        const int check_interval_ms = 5000;
        int elapsed_ms = 0;
        Decimal total_value;

        while (remaining > Decimal::zero() &&
               elapsed_ms < max_duration_.count() * 1000) {
            if (cancelled_.load(std::memory_order_acquire)) {
                result.error = "Cancelled";
                break;
            }

            try {
                auto ticker = client_.ticker(symbol_).get();

                // Check price limit
                if (price_limit_) {
                    if (side_ == Side::Buy && ticker.ask && *ticker.ask > *price_limit_) {
                        std::this_thread::sleep_for(std::chrono::milliseconds(check_interval_ms));
                        elapsed_ms += check_interval_ms;
                        continue;
                    }
                    if (side_ == Side::Sell && ticker.bid && *ticker.bid < *price_limit_) {
                        std::this_thread::sleep_for(std::chrono::milliseconds(check_interval_ms));
                        elapsed_ms += check_interval_ms;
                        continue;
                    }
                }

                Decimal volume = ticker.volume_24h.value_or(Decimal::from_double(1000));
                Decimal interval_volume = volume / Decimal::from_double(24 * 3600000.0 / check_interval_ms);
                Decimal target_qty = interval_volume * target_participation_;
                Decimal qty = (remaining < target_qty) ? remaining : target_qty;

                if (qty > Decimal::zero()) {
                    Order order;
                    if (side_ == Side::Buy) {
                        order = client_.buy(symbol_, qty).get();
                    } else {
                        order = client_.sell(symbol_, qty).get();
                    }

                    result.orders.push_back(order);
                    result.total_filled = result.total_filled + order.filled_quantity;
                    remaining = remaining - order.filled_quantity;

                    if (order.average_price) {
                        total_value = total_value +
                            (order.filled_quantity * *order.average_price);
                    }

                    if (callback_) {
                        callback_(order, remaining);
                    }
                }
            } catch (const std::exception&) {
                // Continue
            }

            std::this_thread::sleep_for(std::chrono::milliseconds(check_interval_ms));
            elapsed_ms += check_interval_ms;
        }

        result.end_time = now_ms();
        result.completed = result.error.empty() && remaining <= Decimal::zero();

        if (result.total_filled.is_positive()) {
            result.average_price = total_value / result.total_filled;
        }

        return result;
    });
}

void PovExecutor::cancel() {
    cancelled_.store(true, std::memory_order_release);
}

}  // namespace lx::trading

File: sdk/lx-trading-ts/src/execution.ts
/**
 * Execution algorithms.
 */

import { Decimal } from 'decimal.js';
import type { Client } from './client.js';
import { orderIsOpen, Side, type Order } from './types.js';

/**
 * Sleep utility.
 */
function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// =============================================================================
// TWAP Executor
// =============================================================================

/**
 * Time-Weighted Average Price execution.
 *
 * Splits a large order into smaller slices executed at regular intervals.
 */
export class TwapExecutor {
  private readonly sliceQty: Decimal;
  private readonly intervalMs: number;

  constructor(
    private readonly client: Client,
    private readonly symbol: string,
    private readonly side: Side,
    private readonly totalQuantity: Decimal,
    durationSeconds: number,
    private readonly numSlices: number,
  ) {
    this.sliceQty = totalQuantity.div(numSlices);
    this.intervalMs = (durationSeconds / numSlices) * 1000;
  }

  /**
   * Execute TWAP strategy.
   */
  async execute(): Promise<Order[]> {
    const orders: Order[] = [];

    for (let i = 0; i < this.numSlices; i++) {
      const remaining = this.totalQuantity.minus(this.sliceQty.mul(i));
      const qty = Decimal.min(this.sliceQty, remaining);

      if (qty.lte(0)) {
        break;
      }

      const order =
        this.side === Side.BUY
          ? await this.client.buy(this.symbol, qty)
          : await this.client.sell(this.symbol, qty);

      orders.push(order);

      if (i < this.numSlices - 1) {
        await sleep(this.intervalMs);
      }
    }

    return orders;
  }
}

// =============================================================================
// VWAP Executor
// =============================================================================

/**
 * Volume-Weighted Average Price execution.
 *
 * Executes based on participation rate of market volume.
 */
export class VwapExecutor {
  private readonly checkIntervalMs = 5000;

  constructor(
    private readonly client: Client,
    private readonly symbol: string,
    private readonly side: Side,
    private readonly totalQuantity: Decimal,
    private readonly participationRate: Decimal, // e.g., 0.1 = 10% of volume
    private readonly maxDurationSeconds: number,
  ) {}

  /**
   * Execute VWAP strategy.
   */
  async execute(): Promise<Order[]> {
    const orders: Order[] = [];
    let remaining = this.totalQuantity;
    let elapsed = 0;

    while (remaining.gt(0) && elapsed < this.maxDurationSeconds * 1000) {
      const ticker = await this.client.ticker(this.symbol);
      const volume = ticker.volume24h ?? new Decimal(1000);

      // Calculate slice based on participation
      const hourlyVolume = volume.div(24);
      const sliceVolume = hourlyVolume
        .mul(this.participationRate)
        .div(3600000 / this.checkIntervalMs);
      const qty = Decimal.min(sliceVolume, remaining);

      if (qty.gt(0)) {
        const order =
          this.side === Side.BUY
            ? await this.client.buy(this.symbol, qty)
            : await this.client.sell(this.symbol, qty);

        orders.push(order);
        remaining = remaining.minus(qty);
      }

      await sleep(this.checkIntervalMs);
      elapsed += this.checkIntervalMs;
    }

    return orders;
  }
}

// =============================================================================
// Iceberg Executor
// =============================================================================

/**
 * Iceberg order execution.
 *
 * Hides large orders by only showing a small visible quantity at a time.
 */
export class IcebergExecutor {
  constructor(
    private readonly client: Client,
    private readonly symbol: string,
    private readonly side: Side,
    private readonly totalQuantity: Decimal,
    private readonly visibleQuantity: Decimal,
    private readonly price: Decimal,
    private readonly venue?: string,
  ) {}

  /**
   * Execute iceberg strategy.
   */
  async execute(): Promise<Order[]> {
    const orders: Order[] = [];
    let remaining = this.totalQuantity;

    while (remaining.gt(0)) {
      const qty = Decimal.min(this.visibleQuantity, remaining);

      const order =
        this.side === Side.BUY
          ? await this.client.limitBuy(this.symbol, qty, this.price, this.venue)
          : await this.client.limitSell(this.symbol, qty, this.price, this.venue);

      // Wait for fill
      let attempts = 0;
      while (orderIsOpen(order) && attempts < 100) {
        await sleep(500);
        attempts++;
        // Note: In production, refresh order status from venue
        break; // Simplified - just continue
      }

      remaining = remaining.minus(order.filledQuantity);
      orders.push(order);
    }

    return orders;
  }
}

// =============================================================================
// Sniper Executor
// =============================================================================

/**
 * Sniper execution - wait for price target then execute immediately.
 */
export class SniperExecutor {
  private readonly checkIntervalMs = 100; // 100ms

  constructor(
    private readonly client: Client,
    private readonly symbol: string,
    private readonly side: Side,
    private readonly quantity: Decimal,
    private readonly targetPrice: Decimal,
    private readonly timeoutSeconds: number,
  ) {}

  /**
   * Execute sniper strategy.
   * Returns the executed order, or undefined if timeout reached.
   */
  async execute(): Promise<Order | undefined> {
    let elapsed = 0;
    const timeoutMs = this.timeoutSeconds * 1000;

    while (elapsed < timeoutMs) {
      const ticker = await this.client.ticker(this.symbol);

      let shouldExecute = false;
      if (this.side === Side.BUY) {
        if (ticker.ask && ticker.ask.lte(this.targetPrice)) {
          shouldExecute = true;
        }
      } else {
        if (ticker.bid && ticker.bid.gte(this.targetPrice)) {
          shouldExecute = true;
        }
      }

      if (shouldExecute) {
        return this.side === Side.BUY
          ? await this.client.buy(this.symbol, this.quantity)
          : await this.client.sell(this.symbol, this.quantity);
      }

      await sleep(this.checkIntervalMs);
      elapsed += this.checkIntervalMs;
    }

    return undefined; // Timeout
  }
}

// =============================================================================
// POV Executor (Percent of Volume)
// =============================================================================

/**
 * Percent of Volume execution.
 *
 * Executes as a percentage of market volume over time.
 */
export class PovExecutor {
  private readonly checkIntervalMs = 10000; // 10 seconds

  constructor(
    private readonly client: Client,
    private readonly symbol: string,
    private readonly side: Side,
    private readonly totalQuantity: Decimal,
    private readonly targetPercent: Decimal, // e.g., 0.05 = 5% of volume
    private readonly maxDurationSeconds: number,
  ) {}

  /**
   * Execute POV strategy.
   */
  async execute(): Promise<Order[]> {
    const orders: Order[] = [];
    let remaining = this.totalQuantity;
    let elapsed = 0;
    let prevVolume = new Decimal(0);

    // Get initial volume
    const initialTicker = await this.client.ticker(this.symbol);
    prevVolume = initialTicker.volume24h ?? new Decimal(0);

    while (remaining.gt(0) && elapsed < this.maxDurationSeconds * 1000) {
      await sleep(this.checkIntervalMs);
      elapsed += this.checkIntervalMs;

      const ticker = await this.client.ticker(this.symbol);
      const currentVolume = ticker.volume24h ?? new Decimal(0);

      // Estimate volume in this interval (rough approximation)
      const volumeDelta = currentVolume.minus(prevVolume).abs();
      prevVolume = currentVolume;

      // Calculate our share
      const ourShare = volumeDelta.mul(this.targetPercent);
      const qty = Decimal.min(ourShare, remaining);

      if (qty.gt(new Decimal('0.0001'))) {
        const order =
          this.side === Side.BUY
            ? await this.client.buy(this.symbol, qty)
            : await this.client.sell(this.symbol, qty);

        orders.push(order);
        remaining = remaining.minus(order.filledQuantity);
      }
    }

    return orders;
  }
}

// =============================================================================
// DCA Executor (Dollar Cost Averaging)
// =============================================================================

/**
 * Dollar Cost Averaging execution.
 *
 * Invests fixed amounts at regular intervals.
 */
export class DcaExecutor {
  constructor(
    private readonly client: Client,
    private readonly symbol: string,
    private readonly side: Side,
    private readonly amountPerInterval: Decimal, // Amount in quote currency
    private readonly intervalMs: number,
    private readonly numIntervals: number,
    private readonly venue?: string,
  ) {}

  /**
   * Execute DCA strategy.
   */
  async execute(): Promise<Order[]> {
    const orders: Order[] = [];

    for (let i = 0; i < this.numIntervals; i++) {
      // Get current price to calculate quantity
      const ticker = await this.client.ticker(this.symbol, this.venue);
      const price = ticker.last ?? ticker.ask ?? ticker.bid;

      if (!price || price.isZero()) {
        continue;
      }

      const qty = this.amountPerInterval.div(price);

      const order =
        this.side === Side.BUY
          ? await this.client.buy(this.symbol, qty, this.venue)
          : await this.client.sell(this.symbol, qty, this.venue);

      orders.push(order);

      if (i < this.numIntervals - 1) {
        await sleep(this.intervalMs);
      }
    }

    return orders;
  }
}
